fuzz: Add fuzz target for block index tree and related validation events

mzumsande opened the pull request:
This adds a fuzz target for the block index and various events in validation that interact with it.

It can create arbitrary tree-like structure of block indexes, simulating (so far) the following events:
- Adding a header
- Receiving the full block (may be valid or not)
- `ActivateBestChain()` - Reorging the chain to a new chain tip (possibly encountering invalid blocks on the way)
- Pruning a block in the best chain
- Receiving a previously pruned block again (`getblockfrompeer`)

It might be interesting / possible to extend this to more events, such as dealing with more than one chainstate (assumeutxo).

The test skips all actual validation of header/ block / transaction data by just simulating the outcome, and also doesn't interact with the data directory.
The main goal is to ensure the integrity of the block index tree in all fuzzed constellations, by calling `CheckBlockIndex()` at the end of each iteration.

Compared to #29158 this approach has a more limited scope (by skipping all actual validation), but it is fast - it doesn't do a full init sequence on each iteration, but "cleans up" after itself by resetting the global validation state after each iteration.

mzumsande committed:
fuzz: Add fuzzer for block index

This fuzz target creates arbitrary tree-like structure of indices,
simulating the following events:
- Adding a header to the block tree db
- Receiving the full block (may be valid or not)
- Reorging to a new chain tip (possibly encountering invalid blocks on
  the way)
- pruning
The test skips all actual validation of header/ block / transaction data
by just simulating the outcome, and also doesn't interact with the data directory.

The main goal is to test the integrity of the block index tree in
all fuzzed constellations, by calling CheckBlockIndex()
at the end of each iteration.

mzumsande commented on src/test/fuzz/block_index_tree.cpp:
```diff
@@ -0,0 +1,201 @@
+            [&] {
+                // Receive a header building on an existing one. This assumes headers are valid, so PoW is not relevant here.
+                LOCK(cs_main);
+                CBlockIndex* prev_block = PickValue(fuzzed_data_provider, blocks);
```
Good point - fixed, and also cleared `blocks` in the end!
While working /testing this branch I directly did picked a value from blockman.m_block_index and introduced `blocks` right before pushing because picking from a `std::unordered_map` mad runs non-deterministic.

ismaelsadeeq commented on src/test/fuzz/block_index_tree.cpp:
```diff
@@ -0,0 +1,203 @@
+            [&] {
+                // Prune chain - dealing with block files is beyond the scope of this test, so just prune random blocks, making no assumptions what must
+                // be together in a block file.
+                // Also don't prune blocks outside of the chain for now - this would make the fuzzer crash because of the problem describted in
```
Maybe lil too early, but a nit when updating.
```suggestion
                // Also don't prune blocks outside of the chain for now - this would make the fuzzer crash because of the problem described in
```

ismaelsadeeq reviewed (commented):
Concept ACK

Simple implementation, comments below are just a suggestions for minor improvements

Crypt-iQ reviewed (commented):
Concept ACK, will test out.

I think both approaches (this PR and the i/o-based https://github.com/bitcoin/bitcoin/pull/29158) have their merits and could both be merged. I like that this PR focuses on the internal state we want to check. I also like that the other PR does round-trip reads/writes to disk since I think there should be asserts on i/o like this.

Crypt-iQ commented:
Have you seen the below crash before / is it the same root cause as reported in #31512?

<details closed>

```
Running: crash-3b11c210b246fa09b395166b5007c451399115b8
Assertion failed: (c->setBlockIndexCandidates.contains(const_cast<CBlockIndex*>(pindex))), function CheckBlockIndex, file validation.cpp, line 5416.
==90212== ERROR: libFuzzer: deadly signal
    #0 0x10876cd70 in __sanitizer_print_stack_trace+0x28 (libclang_rt.asan_osx_dynamic.dylib:arm64+0x5cd70)
    #1 0x105276794 in fuzzer::PrintStackTrace() FuzzerUtil.cpp:210
    #2 0x10525c57c in fuzzer::Fuzzer::CrashCallback() FuzzerLoop.cpp:231
    #3 0x19b787620 in _sigtramp+0x34 (libsystem_platform.dylib:arm64+0x3620)
    #4 0x19b74d888 in pthread_kill+0x124 (libsystem_pthread.dylib:arm64+0x6888)
    #5 0x19b656c5c in abort+0x78 (libsystem_c.dylib:arm64+0x78c5c)
    #6 0x19b655ee8 in __assert_rtn+0x118 (libsystem_c.dylib:arm64+0x77ee8)
    #7 0x10506d1e4 in ChainstateManager::CheckBlockIndex() const validation.cpp:5416
    #8 0x102c40e84 in block_index_tree_fuzz_target(std::__1::span<unsigned char const, 18446744073709551615ul>) block_index_tree.cpp:173
    #9 0x1035242fc in LLVMFuzzerTestOneInput fuzz.cpp:215
    #10 0x10525d9a4 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) FuzzerLoop.cpp:614
    #11 0x10524a6e8 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) FuzzerDriver.cpp:327
    #12 0x10524f980 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) FuzzerDriver.cpp:862
    #13 0x10527707c in main FuzzerMain.cpp:20
    #14 0x19b3aeb48  (<unknown module>)

```
</details>

mzumsande commented:
> Have you seen the below crash before / is it the same root cause as reported in #31512?

no, I haven't seen it, and it shouldn't be the same root cause - could you provide me the seed?

Crypt-iQ commented on src/test/fuzz/block_index_tree.cpp:
```diff
@@ -0,0 +1,205 @@
+            [&] {
+                // Prune chain - dealing with block files is beyond the scope of this test, so just prune random blocks, making no assumptions what must
+                // be together in a block file.
+                // Also don't prune blocks outside of the chain for now - this would make the fuzzer crash because of the problem described in
```
Could there also be a lambda that downloads a previously valid, pruned block that is in the main chain?

Crypt-iQ replied:
> you mean to simulate getblockfrompeer?

Yup just to see that it's tolerated properly. I completely agree that `getblockfrompeer` is kind of a hack.

Crypt-iQ reviewed (commented):
It would be nice if we could also test out `LoadBlockIndex` since it inserts into `m_blocks_unlinked` for pruned blocks, but I think that will require a bit more mocking and not necessary for this PR.

There is some non-determinism in the fuzz test that I'll investigate:
<details /closed>

```
diff --git a/Users/eugenesiegel/btc/bitcoin-review/build/fuzz_det_cov.show.t0.a.txt b/Users/eugenesiegel/btc/bitcoin-review/build/fuzz_det_cov.show.t0.b.txt
index 45c6f3edb4..592f5964eb 100644
--- a/Users/eugenesiegel/btc/bitcoin-review/build/fuzz_det_cov.show.t0.a.txt
+++ b/Users/eugenesiegel/btc/bitcoin-review/build/fuzz_det_cov.show.t0.b.txt
@@ -298059,13 +298059,13 @@
  5513|  19.6k|            } else if (pindexPar == best_hdr_chain[nHeight - 1]) {
                                  ^15.6k^15.6k
   ------------------
-  |  Branch (5513:24): [True: 6.04k, False: 9.57k]
+  |  Branch (5513:24): [True: 6.03k, False: 9.57k]
   ------------------
  5514|       |                // Move to pindex's sibling on the best-chain, if it has one.
- 5515|  6.04k|                pindex = best_hdr_chain[nHeight];
+ 5515|  6.03k|                pindex = best_hdr_chain[nHeight];
  5516|       |                // There will not be a next block if (and only if) parent block is the best header.
- 5517|  6.04k|                assert((pindex == nullptr) == (pindexPar == best_hdr_chain.Tip()));
- 5518|  6.04k|                break;
+ 5517|  6.03k|                assert((pindex == nullptr) == (pindexPar == best_hdr_chain.Tip()));
+ 5518|  6.03k|                break;
  5519|  9.57k|            } else {
  5520|       |                // Move up further.
  5521|  9.57k|                pindex = pindexPar;
@@ -299753,10 +299753,10 @@
                                                                          ^6.06k
   ------------------
   |  Branch (6425:13): [True: 6.06k, False: 5.70k]
-  |  Branch (6425:60): [True: 647, False: 5.41k]
+  |  Branch (6425:60): [True: 645, False: 5.42k]
   ------------------
- 6426|    647|            m_best_header = &entry.second;
- 6427|    647|        }
+ 6426|    645|            m_best_header = &entry.second;
+ 6427|    645|        }
  6428|  11.7k|    }
  6429|  1.98k|}
  6430|       |
⚠️

The coverage was not deterministic between runs.
```
</details>

Crypt-iQ reviewed (commented):
I tracked down the `m_best_header` non-determinism I [posted above](https://github.com/bitcoin/bitcoin/pull/31533#pullrequestreview-2950384616) to `RecalculateBestHeader`. It seems that since `m_block_index` is a `std::unordered_map`, the iteration order in `RecalculateBestHeader` is not guaranteed.

I narrowed down my corpus to two files that, when run in separate processes (the first step of the `deterministic-fuzz-coverage` script), are deterministic. When they are run in the same process (the second step of the script), it is non-deterministic. I found this pretty confusing because `m_block_index` is cleared between runs (besides genesis). It seems that depending on the execution order, `m_block_index` may have changed internal map state (bucketing?) that is not reset by the calls to `erase` in this fuzz test. When I change the map to be a `std::map`, the non-determinism goes away. I am not sure if this can be fixed in this PR?

mzumsande commented:
> It seems that since m_block_index is a std::unordered_map, the iteration order in RecalculateBestHeader is not guaranteed.

Oh yes, this bit me in an earlier iteration of this PR where I made the mistake to do `PickValue(fuzzed_data_provider, blockman.m_block_index)` directly to pick a random block index (so that no `blocks` vector was needed).

It's also related to the fact that we don't use `CBlockIndexWorkComparator` tiebreak rules for `m_best_header` (documented in #32843), so I think it's currently intrinsically non-deterministic (not just in the context of the fuzz test) which block is the best header in case of multiple candidates with the same work?!

sedited reviewed (approved):
ACK 6abbb6ed6cb29b20645a7ef78bcf2e51fc7bde49

Crypt-iQ commented on src/test/fuzz/block_index_tree.cpp:
```diff
@@ -0,0 +1,206 @@
+        genesis->nStatus |= BLOCK_HAVE_UNDO;
+        chainman.m_best_header = genesis;
+        chainman.m_best_invalid = nullptr;
+        chainman.nBlockSequenceId = 1;
```
Should this be 2 instead? See [here](https://github.com/bitcoin/bitcoin/blob/9a29b2d331eed5b4cbd6922f63e397b68ff12447/src/validation.h#L1058).

Crypt-iQ reviewed (commented):
Looks good, left some comments.

mzumsande committed:
test: Wrap validation functions with TestChainstateManager

This allows to access them in the fuzz test in the next commit
without making them public.

Co-authored-by: TheCharlatan <seb.kung@gmail.com>

Crypt-iQ committed:
fuzz: add subtest for re-downloading a previously pruned block

This imitates the use of the getblockfrompeer rpc.
Note that currently pruning is limited to blocks in the active chain.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

sedited reviewed (approved):
re-ACK ac67ab76470441ed7539eb780efa3912f1281faa

The addition of the re-download case is nice.

Crypt-iQ reviewed (commented):
crACK ac67ab76470441ed7539eb780efa3912f1281faa

Will continue to fuzz and see if anything pops up.

maflcko reviewed (commented):
lgtm, but I haven't reviewed the ABC part and I wonder why it is needed.

review ACK db2d39f642979f929261e5f1cd67f0c2f2ca045f 🍶

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK db2d39f642979f929261e5f1cd67f0c2f2ca045f 🍶
KY7sn/dZsxGI5nYXf4tSZsmdUGBzYaUOUVbr+K87FJ+Dcko8OdPvto/RQhmORes6D400eQgSy4bzub8iPcvGCg==
```

</details>

sedited reviewed (approved):
Re-ACK db2d39f642979f929261e5f1cd67f0c2f2ca045f

mzumsande commented on src/test/fuzz/block_index_tree.cpp:
```diff
@@ -0,0 +1,206 @@
+                            if (fuzzed_data_provider.ConsumeBool()) { // Invalid
+                                BlockValidationState state;
+                                state.Invalid(BlockValidationResult::BLOCK_CONSENSUS, "consensus-invalid");
+                                chainman.ActiveChainstate().InvalidBlockFound(block, state);
```
see #34254
